plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/SourceRole.kt

```diff
@@ -28,6 +29,16 @@ sealed interface SourceRole {
         fun ErrorReason.fetchLeakingFunction(): FirBasedSymbol<*> =
             extractInfoFromFunctionArgument(0).unwrap<FirSymbolHolder>().firSymbol
     }
+
+    data class ConditionalEffect(val lhs: SourceRole?, val rhs: SourceRole?) : SourceRole
```

Comment:
Why are these source roles optional?

If you want to be consistent with the terminology of the KEEP, you may want to call these `effect` and `condition`.

Perhaps it would make sense to provide interfaces specifically for effects and conditions, by the way, so that you have less pattern matching to do.

Reply:
> Why are these source roles optional?
Because the `ExpEmbedding`'s `SourceRole` field is optional. When we visit a condition, or an effect, the returned type is the most generic one.

> If you want to be consistent with the terminology of the KEEP, you may want to call these effect and condition. Perhaps it would make sense to provide interfaces specifically for effects and conditions, by the way, so that you have less pattern matching to do.

Good idea, thank you for the feedback Anton!

Reply:
But the code downstream will require these SourceRoles, won't it?  So if they are null we should identify this as an error as early as possible.